paddle/fluid/operators/math/blas_impl.h

```diff
@@ -690,51 +725,93 @@ void Blas<DeviceContext>::MatMul(const framework::Tensor &mat_a,
+                               dim_b.width_ / head_number, dim_b.height_, alpha,
+                               mat_a.data<T>() + sub_matA_offset, lda,
+                               mat_b.data<T>() + sub_matB_offset, ldb, beta,
+                               mat_out->data<T>() + sub_matC_offset, ldc);
+      }
+    } else {
+      PADDLE_ENFORCE_EQ((dim_a.batch_size_ == dim_b.batch_size_ ||
+                         dim_a.batch_size_ == 0 || dim_b.batch_size_ == 0),
+                        true);
+
+      this->template BatchedGEMMWithHead<T>(
+          transA, transB, dim_a.width_, dim_a.height_, dim_b.width_,
+          dim_b.height_, alpha, mat_a.data<T>(), mat_b.data<T>(), beta,
+          mat_out->data<T>(),
+          dim_a.batch_size_ == 0 ? dim_b.batch_size_ : dim_a.batch_size_,
+          dim_a.stride_, dim_b.stride_, head_number, mat_b_split_vertical);
     }
-  } else {
-    PADDLE_ENFORCE(dim_a.batch_size_ == dim_b.batch_size_ ||
-                   dim_a.batch_size_ == 0 || dim_b.batch_size_ == 0);
 
-    this->template BatchedGEMMWithHead<T>(
-        transA, transB, dim_a.height_, dim_b.width_, dim_a.width_, alpha,
-        mat_a.data<T>(), mat_b.data<T>(), beta, mat_out->data<T>(),
-        dim_a.batch_size_ == 0 ? dim_b.batch_size_ : dim_a.batch_size_,
-        dim_a.stride_, dim_b.stride_, head_number);
+  } else {
+    if (dim_a.batch_size_ == 0 && dim_b.batch_size_ == 0) {
+      for (int i = 0; i < head_number; i++) {
+        int sub_matA_offset =
+            dim_a.trans_ ? i * (dim_a.width_ / head_number) * dim_a.height_
+                         : i * (dim_a.width_ / head_number);
+        int sub_matB_offset =
+            dim_b.trans_ ? i * (dim_b.height_ / head_number)
+                         : i * (dim_b.height_ / head_number) * dim_b.width_;
+        int sub_matC_offset = i * dim_b.width_;
+        int lda = !dim_a.trans_ ? dim_a.width_ : dim_a.height_;
+        int ldb = !dim_b.trans_ ? dim_b.width_ : dim_b.height_;
+        int ldc = head_number * dim_b.width_;
+
+        this->template GEMM<T>(transA, transB, dim_a.height_, dim_b.width_,
+                               dim_a.width_ / head_number, alpha,
+                               mat_a.data<T>() + sub_matA_offset, lda,
+                               mat_b.data<T>() + sub_matB_offset, ldb, beta,
+                               mat_out->data<T>() + sub_matC_offset, ldc);
+      }
+    } else {
+      PADDLE_ENFORCE_EQ((dim_a.batch_size_ == dim_b.batch_size_ ||
+                         dim_a.batch_size_ == 0 || dim_b.batch_size_ == 0),
+                        true);
+
+      this->template BatchedGEMMWithHead<T>(
+          transA, transB, dim_a.width_, dim_a.height_, dim_b.width_,
+          dim_b.height_, alpha, mat_a.data<T>(), mat_b.data<T>(), beta,
+          mat_out->data<T>(),
+          dim_a.batch_size_ == 0 ? dim_b.batch_size_ : dim_a.batch_size_,
+          dim_a.stride_, dim_b.stride_, head_number, mat_b_split_vertical);
+    }
   }
```

Comment:
can this part code be refined to avoid duplication?

Reply:
will refactor the code to remove the duplication in the new submit.

Reply:
done